ipa-core/src/protocol/ipa_prf/malicious_security/verifier.rs

```diff
@@ -10,6 +10,43 @@ use crate::{
     },
 };
 
+/// This function computes the shares that sum to zero from the zero-knowledge proofs.
+pub fn compute_g_differences<F, const P: usize, const λ: usize>(
+    zkp: &Vec<[F; P]>,
+    challenges: &[F],
+) -> Vec<F>
+where
+    F: PrimeField,
+{
+    // compute denominator
+    let lagrange_denominator: CanonicalLagrangeDenominator<F, P> =
+        CanonicalLagrangeDenominator::<F, P>::new();
+
+    // compute g_r with "empty" first spot
+    let g_r = iter::once(F::ZERO)
+        .chain(
+            challenges
+                .iter()
+                .zip(zkp)
+                .map(|(challenge, zkp)| interpolate_at_r(zkp, *challenge, &lagrange_denominator)),
```

Comment:
So after we chain together -m/2 and the random challenge points, let's call the loop variable something like expected_sum.